ambry-frontend/src/main/java/com/github/ambry/frontend/HeadBlobHandler.java

```diff
@@ -0,0 +1,172 @@
+    this.accountAndContainerInjector = accountAndContainerInjector;
+    this.metrics = metrics;
+    this.clusterMap = clusterMap;
+    this.quotaManager = quotaManager;
+  }
+
+  void handle(RestRequest restRequest, RestResponseChannel restResponseChannel, Callback<Void> callback)
+      throws RestServiceException {
+    // named blob requests have their account/container in the URI, so checks can be done prior to ID conversion.
+    if (getRequestPath(restRequest).matchesOperation(Operations.NAMED_BLOB)) {
+      accountAndContainerInjector.injectAccountAndContainerForNamedBlob(restRequest, metrics.headBlobMetricsGroup);
+    }
+    new CallbackChain(restRequest, restResponseChannel, callback).start();
+  }
+
+  private class CallbackChain {
+    private final RestRequest restRequest;
+    private final RestResponseChannel restResponseChannel;
+    private final Callback<Void> finalCallback;
+
+    /**
+     * @param restRequest the {@link RestRequest}.
+     * @param restResponseChannel the {@link RestResponseChannel}.
+     * @param finalCallback the {@link Callback} to call on completion.
+     */
+    private CallbackChain(RestRequest restRequest, RestResponseChannel restResponseChannel,
+        Callback<Void> finalCallback) {
+      this.restRequest = restRequest;
+      this.restResponseChannel = restResponseChannel;
+      this.finalCallback = finalCallback;
+    }
+
+    /**
+     * Start the chain by calling {@link SecurityService#processRequest}.
+     */
+    private void start() {
+      restRequest.setArg(RestUtils.InternalKeys.KEEP_ALIVE_ON_ERROR_HINT, true);
+      securityService.processRequest(restRequest, securityProcessRequestCallback());
+    }
+
+    /**
+     * After {@link SecurityService#processRequest} finishes, call {@link IdConverter#convert} to convert the incoming
+     * ID if required.
+     * @return a {@link Callback} to be used with {@link SecurityService#processRequest}.
+     */
+    private Callback<Void> securityProcessRequestCallback() {
+      return buildCallback(metrics.headBlobSecurityProcessRequestMetrics, result -> {
+        String blobIdStr = getRequestPath(restRequest).getOperationOrBlobId(false);
+        idConverter.convert(restRequest, blobIdStr, idConverterCallback());
+      }, restRequest.getUri(), LOGGER, finalCallback);
+    }
+
+    /**
+     * After {@link IdConverter#convert} finishes, call {@link SecurityService#postProcessRequest} to perform
+     * request time security checks that rely on the request being fully parsed and any additional arguments set.
+     * @return a {@link Callback} to be used with {@link IdConverter#convert}.
+     */
+    private Callback<String> idConverterCallback() {
+      return buildCallback(metrics.headBlobIdConversionMetrics, convertedBlobId -> {
+        BlobId blobId = FrontendUtils.getBlobIdFromString(convertedBlobId, clusterMap);
+        accountAndContainerInjector.injectTargetAccountAndContainerFromBlobId(blobId, restRequest,
```

Comment:
May be missing check to see if the Account and container is not already injected (for named blobs)?

Reply:
updated.